pkg/driver/state.go

```diff
@@ -16,24 +16,34 @@ const (
 	// StateRunning means that the driver has been sending data. The caller
 	// who started the driver may start reading data from the hardware.
 	StateRunning = "running"
+	// StateGood means the driver is available for use by applications.
```

Comment:
you mentioned that only linux returns this now but this means that end users need to understand all of these states. is there a way to use all of the states or merge the new ones together? I'm concerned an end user may not know how to interpret the state.

Reply:
Got it. Let's see if we can work with this to keep the existing enums.
1. This would be StateClosed; nothing in the process has opened it yet
2. This would be StateRunning as you already pointed out
3. This would be StateClosed if we consider closed to be the default
4. This would be StateClosed if we consider that if we cannot find something, it is not open

If we did things this way, could the status of camera linux use its current state plus these cases to produce the right value?

Reply:
I don't think so given the ambiguity of open and closed for a v4l2 driver _but_ I think there's a better way. Since the big thing we want to avoid here is breaking code for current users would versioning these changes work?

These changes are to support exposing the experimental `Initialize` function since calling that function ruins the state logic we have in place. For current users, neither this change nor the `Initialize()` function is needed. We could put these changes in `pion/mediadevices/tree/master/pkg/experimental` or a `v2` folder and merge them all at once in a future version once they're more stable. I _would_ suggest adding these changes to a different major version but since we don't have one yet, and as far as I've seen there is no current consensus on what the first version of this software would look like, it's kind of a moot point.

Reply:
Why wouldn't that work given the ambiguity? What breaks upstream? I' m not sure about adding a v2/experimental for this if we can still make something work

Reply:
You do raise a good point that this isn't 1.0 yet but I'm still curious if we can make this work with what we have an addition to @at-wat's feedback

Reply:
> Why wouldn't that work given the ambiguity?

What exactly does an _open_ and _closed_ state mean or imply for a device driver? In the truest sense it means something has an open file descriptor to it. It implies nothing in particular. For example, if the user has a device at `/dev/video0` we can see what processes opened this file 
```
$ lsof -t /dev/video0
46915
...
$ ps -ef | grep myApp  
$USER      46915   46729  7 15:14 pts/4    00:00:18 myApp
```  
If it's _closed_ then `lsof` will have an empty process list and no output. What does this imply or in other words, what can (or can't) we do based off of this information? To get the status of the driver we need to open the driver, query its status, then close the driver. The driver may or may not have already been opened or closed. If we continue to assume that drivers can only be opened once (see [toOpened](https://github.com/pion/mediadevices/blob/bf290b026c1037ba0ff839a41d42d86f666d4455/pkg/driver/state.go#L45)) we will never be able to get the status of an opened driver because we won't be able to open it. 

>  1. This would be StateClosed; nothing in the process has opened it yet

We call `Initialize` and discover what we think is a new driver. Have we opened it yet? We would need to map the `camera.path` to a boolean indicating whether we called `Open`. This ignores the fact that another process may have opened the file and I'm not sure who this info is useful for per my comment above.

> 3. This would be StateClosed if we consider closed to be the default

There's not a "default" state for a driver that you can't get state information from. See some common error states [here](https://www.kernel.org/doc/html/v4.9/media/uapi/gen-errors.html#id1). Those errors are returned from the (VIDIOC_LOG_STATUS)[https://www.kernel.org/doc/html/v4.9/media/uapi/v4l/vidioc-log-status.html] ioctl. Device drivers aren't guaranteed to implement that ioctl so I used a different set of ioctls but same deal. 

> 4. This would be StateClosed if we consider that if we cannot find something, it is not open

How would a user know if they can attempt to open a driver and start streaming? If both condition 4 and 1 are the same it could mean either, "nothing is streaming from this driver so feel free to do so" or "this driver is not connected so do not stream." It's inherently ambiguous. 

 > What breaks upstream? 
 
 Anyone using the `driver.Status` will now have to update their code based on these new states. It doesn't necessarily break the API in the literal sense but users may have to re-write code or this will introduce subtle runtime errors.

Reply:
Noodling on a thought. Considering windows and macos won't be updated here. Maybe a temporary pre 1.0 thing could be to introduce an adapter specific state that can be optionally used in conjunction with driver state. Then v1 can merge the two or have your way supersede the two